### PR TITLE
fix(Popup|Dialog): ESC should close opened Popup/Dialog if focus is lost and moved to body

### LINKED ISSUE #1709: Dialog: can't close if not focused
### Steps
1. Go to `Dialog` examples: https://stardust-ui.github.io/react/components/dialog
2. Open any `Dialog`
3. Click inside of it, on a header for example
4. Press `Escape` key

### Expected Result
`Dialog` is closed

### Actual Result
`Dialog` is open because focus was moved to `body` while we have an event handler on `Dialog` `root`

### Version
0.34.2

### PR DESCRIPTION
**Pull request by sophieH29**
Fixes https://github.com/stardust-ui/react/issues/1709

### CODE REVIEW

**Comment by lucivpav** [comment]
Can we add a test?

File: packages/react/src/components/Dialog/Dialog.tsx [hunk 1]
```diff
@@ -203,6 +204,20 @@ class Dialog extends AutoControlledComponent<WithAsProp<DialogProps>, DialogStat
     }
   }
 
+  handleDocumentKeydown = (getRefs: Function) => e => {
```
Comment (layershifter) [inline packages/react/src/components/Dialog/Dialog.tsx]:
```suggestion
  handleDocumentKeydown = (getRefs: Function) => (e: KeyboardEvent) => {
```

More type safety 🏋

File: packages/react/src/components/Dialog/Dialog.tsx [hunk 2]
```diff
@@ -203,6 +204,20 @@ class Dialog extends AutoControlledComponent<WithAsProp<DialogProps>, DialogStat
...
+    // if focus was lost from Dialog, for e.g. when click on Dialog's content
+    // and ESC is pressed, the opened Dialog should get closed and the trigger should get focus
+    const lastOverlayRef = getRefs().pop()
+    const isLastOpenedDialog: boolean =
+      lastOverlayRef && lastOverlayRef.current === this.overlayRef.current
+
+    const keyCode = keyboardKey.getCode(e)
+    if (keyCode === keyboardKey.Escape && isLastOpenedDialog && this.state.open) {
```
Comment (layershifter) [inline packages/react/src/components/Dialog/Dialog.tsx]:
```suggestion
    if (keyboardKey.getCode(e) === keyboardKey.Escape && isLastOpenedDialog) {
```

You don't need `this.state.open` as this event listener will be mounted only when `Dialog` is `open`:

```tsx
<Portal open={open} /* here we are */>
  <EventListener type="keydown" />
</Portal>
```

**Review by layershifter — APPROVED** [review]

File: packages/react/src/components/Dialog/Dialog.tsx [hunk 3]
```diff
@@ -312,6 +326,12 @@ class Dialog extends AutoControlledComponent<WithAsProp<DialogProps>, DialogStat
                 type="click"
                 capture
               />
+              <EventListener
+                listener={this.handleDocumentKeydown(getRefs)}
+                targetRef={documentRef}
```
Comment (layershifter) [inline packages/react/src/components/Dialog/Dialog.tsx]:
```suggestion
                targetRef={targetRef}
```

**Pull request merged**